src/error.rs

```diff
@@ -67,13 +98,53 @@ impl Error {
     }
 
     pub(crate) fn response(mut self, response: Response) -> Self {
-        self.response = Some(Box::new(response));
+        self.response = Some(response);
         self
     }
-    pub(crate) fn kind(&self) -> ErrorKind {
+
+    /// The type of this error.
+    ///
+    /// ```
+    /// # ureq::is_test(true);
+    /// let err = ureq::get("http://httpbin.org/status/500")
+    ///     .call().unwrap_err();
+    /// assert_eq!(err.kind(), ureq::ErrorKind::HTTP);
+    /// ```
+    pub fn kind(&self) -> ErrorKind {
         self.kind
     }
 
+    /// For Errors of type HTTP (i.e. those that result from an HTTP status code),
+    /// return the status code of the response. For all other Errors, return 0.
+    ///
+    /// ```
+    /// # ureq::is_test(true);
+    /// let err = ureq::get("http://httpbin.org/status/500")
+    ///     .call().unwrap_err();
+    /// assert_eq!(err.kind(), ureq::ErrorKind::HTTP);
+    /// assert_eq!(err.status(), 500);
+    /// ```
+    pub fn status(&self) -> u16 {
+        match &self.response {
+            Some(response) => response.status(),
+            None => 0,
+        }
+    }
+
+    /// For an Error of type HTTP (i.e. those that result from an HTTP status code),
+    /// turn the error into the underlying Response. For other errors, return None.
+    ///
+    /// ```
+    /// # ureq::is_test(true);
+    /// let err = ureq::get("http://httpbin.org/status/500")
+    ///     .call().unwrap_err();
+    /// assert_eq!(err.kind(), ureq::ErrorKind::HTTP);
+    /// assert_eq!(err.into_response().unwrap().status(), 500);
+    /// ```
```

Comment:
Could we get rid of this Optional? 

I know there are some "early" errors like validating the URL which means constructing a Response could be awkward.

I'm just thinking ergonomics.

Reply:
There are actually a lot of error types for which a response is unavailable: All of the Io ones, DnsFailed, ConnectionFailed, BadStatus, BadHeader, etc. I think you're suggesting that for those cases, if someone called into_response(), we'd construct a fake Response with e.g. status 500. I would prefer to avoid that - it mixes up the error layers more than I'd like.

The optional is indeed a bit unfortunate because it encourages checking the ErrorKind and then doing `.unwrap()`. That's safe, but it requires a little reasoning to see its safety and might cause the user to ignore other instances of `.unwrap()`. That said, short of constructing fake Responses, I don't see a way around it.